hls-test-utils/src/Test/Hls.hs

```diff
@@ -308,6 +391,90 @@ runSessionWithServerFormatter plugin formatter conf fp act = do
     fp
     act
 
+runSessionWithServerInTmpDir :: Pretty b => PluginTestDescriptor b -> VirtualFileTree -> Session a -> IO a
+runSessionWithServerInTmpDir plugin tree act = do
+  recorder <- pluginTestRecorder
+  runSessionWithServerInTmpDir' (plugin recorder) def def fullCaps tree act
+
+runSessionWithServerAndCapsInTmpDir :: Pretty b => PluginTestDescriptor b -> ClientCapabilities -> VirtualFileTree -> Session a -> IO a
+runSessionWithServerAndCapsInTmpDir plugin caps tree act = do
+  recorder <- pluginTestRecorder
+  runSessionWithServerInTmpDir' (plugin recorder) def def caps tree act
+
+runSessionWithServerFormatterInTmpDir :: Pretty b => PluginTestDescriptor b -> String -> PluginConfig ->  VirtualFileTree -> Session a -> IO a
+runSessionWithServerFormatterInTmpDir plugin formatter conf tree act = do
+  recorder <- pluginTestRecorder
+  runSessionWithServerInTmpDir'
+    (plugin recorder)
+    def
+      { formattingProvider = T.pack formatter
+      , plugins = M.singleton (PluginId $ T.pack formatter) conf
+      }
+    def
+    fullCaps
+    tree
+    act
+
+-- | Host a server, and run a test session on it.
+--
+-- Creates a temporary directory, and materializes the VirtualFileTree
+-- in the temporary directory.
+--
+-- To debug test cases and verify the file system is correctly set up,
+-- you should set the environment variable 'HLS_TEST_HARNESS_NO_TESTDIR_CLEANUP=1'.
+-- Further, we log the temporary directory location on startup. To view
+-- the logs, set the environment variable 'HLS_TEST_HARNESS_STDERR=1'.
+--
+-- Example invocation to debug test cases:
+--
+-- @
+--   HLS_TEST_HARNESS_NO_TESTDIR_CLEANUP=1 HLS_TEST_HARNESS_STDERR=1 cabal test <plugin-name>
+-- @
+--
+-- Don't forget to use 'TASTY_PATTERN' to debug only a subset of tests.
+--
+-- For plugin test logs, look at the documentation of 'mkPluginTestDescriptor'.
+--
+-- Note: cwd will be shifted into a temporary directory in @Session a@
+runSessionWithServerInTmpDir' ::
+  -- | Plugins to load on the server.
+  --
+  -- For improved logging, make sure these plugins have been initalised with
+  -- the recorder produced by @pluginTestRecorder@.
+  IdePlugins IdeState ->
+  -- | lsp config for the server
+  Config ->
+  -- | config for the test session
+  SessionConfig ->
+  ClientCapabilities ->
+  VirtualFileTree ->
+  Session a ->
+  IO a
+runSessionWithServerInTmpDir' plugins conf sessConf caps tree act = withLock lockForTempDirs $ do
```

Comment:
At the moment because `runSessionWithServer'` uses `withCurrentDirectory`, or something like that. Which is a global variable. If the integration tests run in parallel, they affect each other's current working directories, and then all fail. I haven't looked into it whether we can remove the `withCurrentDirectory` call...

Reply:
Ah, for some reason I thought we were actually running the HLSs in subprocesses, but we aren't.

Reply:
I think we can, in my other branch that doesn't use lsp-test, I don't need the lock or `withCurrentDirectory`.